x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotRecoveryStateIntegrationTests.java

```diff
@@ -70,18 +90,60 @@ public void testRecoveryStateRecoveredBytesMatchPhysicalCacheState() throws Exce
 
         assertAcked(client().admin().indices().prepareDelete(indexName));
 
-        final MountSearchableSnapshotRequest req = new MountSearchableSnapshotRequest(
-            restoredIndexName,
-            fsRepoName,
-            snapshotInfo.snapshotId().getName(),
+        mountSnapshot(fsRepoName, snapshotName, indexName, restoredIndexName, Settings.EMPTY);
+        ensureGreen(restoredIndexName);
+
+        final Index restoredIndex = client().admin()
+            .cluster()
+            .prepareState()
+            .clear()
+            .setMetadata(true)
+            .get()
+            .getState()
+            .metadata()
+            .index(restoredIndexName)
+            .getIndex();
+
+        assertExecutorIsIdle(SearchableSnapshotsConstants.CACHE_PREWARMING_THREAD_POOL_NAME);
+        assertExecutorIsIdle(SearchableSnapshotsConstants.CACHE_FETCH_ASYNC_THREAD_POOL_NAME);
+
+        RecoveryState recoveryState = getRecoveryState(restoredIndexName);
+
+        assertThat(recoveryState.getStage(), equalTo(RecoveryState.Stage.DONE));
+
+        long recoveredBytes = recoveryState.getIndex().recoveredBytes();
+        long physicalCacheSize = getPhysicalCacheSize(restoredIndex, snapshotInfo.snapshotId().getUUID());
+
+        assertThat("Physical cache size doesn't match with recovery state data", physicalCacheSize, equalTo(recoveredBytes));
+        assertThat("Expected to recover 100% of files", recoveryState.getIndex().recoveredBytesPercent(), equalTo(100.0f));
+    }
+
+    public void testFilesStoredInThePersistentCacheAreMarkedAsReusedInRecoveryState() throws Exception {
+        final String fsRepoName = randomAlphaOfLength(10);
+        final String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+        final String restoredIndexName = randomBoolean() ? indexName : randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+        final String snapshotName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+
+        createRepository(fsRepoName, "test-fs");
+        int numberOfShards = 1;
+
+        createAndPopulateIndex(
             indexName,
-            Settings.EMPTY,
-            Strings.EMPTY_ARRAY,
-            true
+            Settings.builder().put(INDEX_SOFT_DELETES_SETTING.getKey(), true).put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, numberOfShards)
         );
 
-        final RestoreSnapshotResponse restoreSnapshotResponse = client().execute(MountSearchableSnapshotAction.INSTANCE, req).get();
-        assertThat(restoreSnapshotResponse.getRestoreInfo().failedShards(), equalTo(0));
+        final SnapshotInfo snapshotInfo = createFullSnapshot(fsRepoName, snapshotName);
+
+        assertAcked(client().admin().indices().prepareDelete(indexName));
+
+        mountSnapshot(fsRepoName, snapshotName, indexName, restoredIndexName, Settings.EMPTY);
+        ensureGreen(restoredIndexName);
+
+        for (CacheService cacheService : internalCluster().getDataNodeInstances(CacheService.class)) {
```

Comment:
I think we need to wait for the recovery to be `done` before executing `synchronizeCache()`? Otherwise there is a risk that not all cache files are fully prewarmed and synchronized on disk this fails:
> assertThat("Expected to reuse all data from the persistent cache but it didn't", 0L, equalTo(recoveredBytes));

Reply:
Addressed in https://github.com/elastic/elasticsearch/pull/67425/commits/e0c5c8d9eb3b271d669bdedc239ca255b0d510b6